tests/websocket.rs

```diff
@@ -20,6 +38,409 @@ async fn connect_app_websocket() {
+        .get_or_connect_app_client("fixture1".to_string(), admin_ws.clone())
+        .await
+        .unwrap();
+    assert_eq!(
+        "fixture1".to_string(),
+        app_client_1.cached_app_info().installed_app_id
+    );
+
+    let app_client_2 = pool
+        .get_or_connect_app_client("fixture2".to_string(), admin_ws.clone())
+        .await
+        .unwrap();
+    assert_eq!(
+        "fixture2".to_string(),
+        app_client_2.cached_app_info().installed_app_id
+    );
+
+    // Should have provisioned exactly 1 app interface for the HTTP gateway
+    //
+    // Note that this check would also pass if the conductor was exposing an app interface with
+    // allowed origins *.
+    let app_interfaces = sweet_conductor.list_app_interfaces().await.unwrap();
+    let matched_app_interfaces = app_interfaces
+        .iter()
+        .filter(|interface| interface.allowed_origins.is_allowed(HTTP_GW_ORIGIN))
+        .collect::<Vec<_>>();
+    assert_eq!(matched_app_interfaces.len(), 1);
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn reuse_connection() {
+    initialize_tracing_subscriber();
+
+    let sweet_conductor = SweetConductor::from_standard_config().await;
+
+    let app = install_fixture1(sweet_conductor.clone(), None)
+        .await
+        .unwrap();
+    init_zome(sweet_conductor.clone(), &app, "coordinator1".to_string())
+        .await
+        .unwrap();
+
+    let admin_port = sweet_conductor
+        .get_arbitrary_admin_websocket_port()
+        .unwrap();
+    let admin_ws = AdminWebsocket::connect((Ipv4Addr::LOCALHOST, admin_port))
+        .await
+        .unwrap();
+
+    let pool = AppConnPool::new(create_test_configuration(admin_port).unwrap());
+
+    let app_client_1 = pool
+        .get_or_connect_app_client("fixture1".to_string(), admin_ws.clone())
+        .await
+        .unwrap();
+    assert_eq!(
+        "fixture1".to_string(),
+        app_client_1.cached_app_info().installed_app_id
+    );
+
+    // Take out a write lock so that the pool cannot create a new connection
+    let inner_pool = pool.get_inner_pool();
+    let _read_lock = inner_pool.read().await;
+
+    let app_client_1_handle = tokio::time::timeout(std::time::Duration::from_millis(100), {
+        let pool = pool.clone();
+        let admin_ws = admin_ws.clone();
+        async move {
+            pool.get_or_connect_app_client("fixture1".to_string(), admin_ws)
+                .await
+        }
+    })
+    .await
+    .unwrap()
+    .unwrap();
+
+    // Check that the client is usable
+    assert_eq!(
+        "fixture1".to_string(),
+        app_client_1_handle
+            .app_info()
+            .await
+            .unwrap()
+            .unwrap()
+            .installed_app_id
+    );
+
+    // Demonstrate that the pool was prevented from writing by the read lock held above.
+    assert!(inner_pool.try_write().is_err());
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn does_not_reconnect_on_non_websocket_error() {
```

Comment:
I don't understand what this test tests. Would it be possible to add a bit more description beyond what the function name says?

Reply:
Ah okay, because `get_or_connect_app_client` wouldn't be able to get the lock :+1: . I don't understand yet though why the zome call produces an error at all, it looks valid to me.

Reply:
Yes, that warrants a comment.

Reply:
Or perhaps the closure can just return an error instead of making a call?